doc/Requirements.md

```diff
@@ -0,0 +1,526 @@
+
+hardware: system controlled by the PF. Not necessary material system.
+          This term was chosen because:
+ - historically the PF reference implementation was used to abstract hardware
+ - the subsystem term would arguably fit best is already used.
+(FIXME: choose subsystem instead, it is not used in fact.)
+
+## Value
+
+A parameter MUST have a value.
+(Why: because a parameter without value would not abstract any hardware.)
+
+A PF MUST support mutable parameters.
+(Why: To control the underlined hardware.)
+
+A PF MAY support immutable parameters, ie parameters which value is determined on start then read only.
+(Why: To permit hardware read only value reflection.)
+(Note: This is not implemented in the PF reference implementation.)
+
+This value MUST be gettable and settable (except if immutable).
+(Why: A parameter that can not be accessed is of no use.)
+
+A value MUST be convertible to a string.
+(Why: To display a parameter value to the human user for debugging purpose.)
+
+A bijective conversion string <-> value SHOULD exist.
+(Why: To support serialization and deserialization of the parameter values.
+      It is not a MUST because a PF implementation could offer serialization to
+      custom -not string based- format.)
+
+### Data type
+
+#### Definition
+All parameters have a data type.
+A data type designates parameter invariants.
+A data type is the the meaning of the data and the way values of that type can be stored.
+
+
+#### Philosophy
+
+A data type defines the value properties:
+ - memory layout
+ - value constrains
+
+A value type is mostly used to:
+ - pretty display parameter values (not just a as an array of bits)
+ - check for user error when setting it (out of bound, invalid...)
+ - offer a type safe api
+
+#### Requirements
+
+A PF SHOULD support the following types.
+If a type is chosen to be supported, it MUST respect all MUST clause,
+SHOULD respect all SHOULD clause, MAY respect all MAY clause of the type.
+(Why: All type are not necessary to use the PF. For example any parameter could
+      be represented as an array of char (string). But this would not permit to
+      check parameter validity (invariants) as well a pretty displaying the values.)
+
+Implementation MAY add another API to access a parameter value.
+(Why: For example a C++ implementation may give access to a string as an std::string object.)
+
+##### Integers
+PF SHOULD support signed and unsigned integer parameters
+(Why: The reference implementation supports it.)
+
+PF MUST support integer with invariant size.
+(Why: It is common in C API to expect numbers to have a fixed maximum size.)
+
+The API to access it MUST respect C integer abi.
+(Why: For easy access from C code.)
+
+Supported integer size SHOULD be at least 8, 16 and 32 bits.
+(Why: The reference implementation supports it.)
+
+PF MAY support constraining the parameter minimum and maximum value.
+(Why: To catch user out of valid range errors when changing the parameter value.)
```

Comment:
Why minimum and maximum value is only used for integers?
Why not for floating point parameter or fixed point parameter?